pkg/endpoint/endpoint.go

```diff
@@ -2134,23 +2134,16 @@ func (e *Endpoint) Disconnect() string {
 
 // Delete cleans up all resources associated with this endpoint, including the
 // following:
-// * removal from the endpointmanager, resulting in new events not taking effect
-// on this endpoint
 // * cleanup of datapath state (BPF maps, proxy configuration, directories)
 // * releasing IP addresses allocated for the endpoint
 // * releasing of the reference to its allocated security identity
-func (e *Endpoint) Delete(ipam IPReleaser, manager endpointManager, conf DeleteConfig) []error {
+//
+// The caller must first call endpoint.Disconnect() before calling Delete().
```

Comment:
What happens if the caller forgets? Would it make sense to keep the call to `Disconnect()` in here, assuming it's idempotent, so that this contract can be enforced, rather than just advising?

Reply:
Good question, in essence I think this is this point from the PR description:

> To avoid too many changes, this PR does not actually change the ordering of endpoint.Stop() vs. removal from EndpointManager though I could see an argument that the endpoint should be unexposed from other subsystems before halting the event processing goroutines for the endpoint. That can be considered as future work.

Basically in the version of the code that this PR ends up at today, `WaitEndpointRemoved()` does a `Disconnect()` first to stop the endpoint from handling events, then removes the endpoint from the `EndpointManager`, then continues with the rest of the cleanup. The fact that the middle part there is the responsibility of the `EndpointManager` while the `Disconnect()` and `Delete()` are the responsibility of the `Endpoint` package is the main reason why the `Disconnect()` isn't just part of `Delete()`.

That said, I don't have a good answer why not to do as you suggest, other than it changes the order of operations from the current logic flow. In fact, I think your suggestion is the right approach:
* On endpoint create, first start the goroutines to handle events, then expose in endpointmanager
* On endpoint delete, do the reverse: unexpose then stop the goroutines.

The latter point is the wrong way around right now. Maybe I should just bite the bullet and switch it around.

Reply:
Ah good call out, yeah I had forgotten that you mentioned this already. I'm OK either way to fix it up in a followup or to do it here. On second thought, I think I'd lean slightly towards a followup just because ideally a refactor _should_ not change behavior. Up to you ultimately.

Reply:
This is actually a nice cleanup. Fingers crossed nothing falls out from CI ;-)

```
$ git diff HEAD~2
diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index 39d1f79c6399..5d627d3fc069 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -2074,29 +2074,22 @@ type IPReleaser interface {
        ReleaseIP(net.IP) error
 }

-// Disconnect is similar to Stop(), but additionally sets the endpoint into
-// StateDisconnecting which prevents subsequent operations from occurring on
-// the endpoint (governed by lockAlive()).
-//
-// The caller is expected to subsequently call endpoint.Delete() to clean up
-// remaining resources.
-//
-// Returns the state that the endpoint was in prior to this call.
-func (e *Endpoint) Disconnect() string {
-       return e.stop(true)
-}
-
 // Delete cleans up all resources associated with this endpoint, including the
 // following:
 // * cleanup of datapath state (BPF maps, proxy configuration, directories)
 // * releasing IP addresses allocated for the endpoint
 // * releasing of the reference to its allocated security identity
-//
-// The caller must first call endpoint.Disconnect() before calling Delete().
 func (e *Endpoint) Delete(ipam IPReleaser, conf DeleteConfig) []error {
        errs := []error{}

+       e.Stop()
+       if err := e.lockAlive(); err != nil {
+               return errs
+       }
+       defer e.unlock()
+
        e.unconditionalLock()
+       e.setState(StateDisconnecting, "Deleting endpoint")

        // If dry mode is enabled, no changes to BPF maps are performed
        if !option.Config.DryMode {
diff --git a/pkg/endpoint/events.go b/pkg/endpoint/events.go
index d9621ae62ea2..0b16610ad4ae 100644
--- a/pkg/endpoint/events.go
+++ b/pkg/endpoint/events.go
@@ -343,7 +343,7 @@ func (e *Endpoint) InitEventQueue() {
 // receive events from other subsystems.
 //
 // The endpoint must not already be exposed via the endpointmanager prior to
-// calling Expose(), as it assumes unconditional access over the Endpoint
+// calling Start(), as it assumes unconditional access over the Endpoint
 // object.
 func (e *Endpoint) Start(id uint16) {
        // No need to check liveness as an endpoint can only be deleted via the
@@ -366,7 +366,13 @@ func (e *Endpoint) Start(id uint16) {
        e.getLogger().Info("New endpoint")
 }

-func (e *Endpoint) stop(disconnect bool) string {
+// Stop cleans up all goroutines managed by this endpoint (EventQueue,
+// Controllers).
+// This function should be used directly in cleanup functions which aim to stop
+// goroutines managed by this endpoint, but without removing BPF maps and
+// datapath state (for instance, because the daemon is shutting down but the
+// endpoint should remain operational while the daemon is not running).
+func (e *Endpoint) Stop() {
        // Since the endpoint is being deleted, we no longer need to run events
        // in its event queue. This is a no-op if the queue has already been
        // closed elsewhere.
@@ -386,28 +392,6 @@ func (e *Endpoint) stop(disconnect bool) string {
        // enqueued for this endpoint, this is a no-op.
        e.closeBPFProgramChannel()

-       if err := e.lockAlive(); err != nil {
-               return e.GetState()
-       }
-       defer e.unlock()
-
        // Cancel active controllers for the endpoint tied to e.aliveCtx.
        e.aliveCancel()
-       if disconnect {
-               e.setState(StateDisconnecting, "Deleting endpoint")
-       }
-
-       return e.getState()
-}
-
-// Stop cleans up all goroutines managed by this endpoint (EventQueue,
-// Controllers).
-// This function should be used directly in cleanup functions which aim to stop
-// goroutines managed by this endpoint, but without removing BPF maps and
-// datapath state (for instance, because the daemon is shutting down but the
-// endpoint should remain operational while the daemon is not running).
-func (e *Endpoint) Stop() {
-       // Callers of endpoint.Stop() don't modify the state, so there's no
-       // need to return the state to them.
-       _ = e.stop(false)
 }
diff --git a/pkg/endpointmanager/manager.go b/pkg/endpointmanager/manager.go
index 9f5c8a04507f..4bea60897dcc 100644
--- a/pkg/endpointmanager/manager.go
+++ b/pkg/endpointmanager/manager.go
@@ -316,7 +316,7 @@ func (mgr *EndpointManager) WaitEndpointRemoved(ep *endpoint.Endpoint) {
                // Already disconnecting
                return
        }
-       previousState := ep.Disconnect()
+       previousState := ep.GetState()

        // This must be done before the ID is released for the endpoint!
        mgr.removeIDLocked(ep.ID)
```